### PR TITLE
Specialized Collections Tests

### PR DESCRIPTION
**Pull request by hughbe**
Improve code coverage to 100% for System.Collections.Specialized
Remove dead code

### CODE REVIEW

File: src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs [hunk 1]
```diff
@@ -224,7 +224,7 @@ public bool Contains(object key)
         public void CopyTo(Array array, int index)
         {
             if (array == null)
-                throw new ArgumentNullException("array");
+                throw new ArgumentNullException("array", SR.ArgumentNull_Array);
```
Comment (justinvp) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs]: Does `ArgumentNull_Array` add enough value to justify keeping it and using it throughout? I wonder if this resource string should be removed altogether from this assembly. One less string resource to have to embed and localize.

The default message for `new ArgumentNullException("array")` is:

```
Value cannot be null.
Parameter name: array
```

Versus `new ArgumentNullException("array", SR.ArgumentNull_Array)`:

```
Array cannot be null.
Parameter name: array
```

The latter is more precise, but the former seems "good enough" IMHO.

**Comment by Clockwork-Muse** [comment]
Side note: [I've done some work on this off and on](https://github.com/Clockwork-Muse/corefx/tree/516_System.Collections.Specialized), as part of a couple of other things.  Given likelihood of change to this particular library, I wasn't giving it any high priority.  Don't know if anything I've done would be useful, and there's a few things that need to be cleaned up with what I've done (double checking about some of the string-key-comparisons in terms of user default culture comes to mind).
It would be good to, in general:  
 - Clean up the `Assert.True(false, ...)` and `Assert.False(true, ...)` instances for something more appropriate (there's ~1800 of them here)
 - Correct test namespaces.  There are several conflicts to manage, many of which I've already handled.
 - Unstep the "internal" test framework present in some cases - there's usually a better way to manage some of that kind of testing.

I'm not sure what changes will be accepted for this library - my earlier attempt to get in the BitVector32 Debug one was rejected.  Was holding off on the enumerator one in Hybrid until I'd done more work on it (just added the relevant test for now).

File: src/System.Collections.Specialized/src/Resources/Strings.resx [hunk 2]
```diff
@@ -153,16 +153,4 @@
...
-  </data>
-  <data name="Arg_ArrayPlusOffTooSmall" xml:space="preserve">
-    <value>Destination array is not long enough to copy all the items in the collection. Check array index and length.</value>
-  </data>
-  <data name="NotSupported_KeyCollectionSet" xml:space="preserve">
-    <value>Mutating a key collection derived from a dictionary is not allowed.</value>
-  </data>
-  <data name="NotSupported_ValueCollectionSet" xml:space="preserve">
-    <value>Mutating a value collection derived from a dictionary is not allowed.</value>
-  </data>
```
Comment (stephentoub) [inline src/System.Collections.Specialized/src/Resources/Strings.resx]: Can you separate the product changes out into a separate commit?  This is about removing dead code, right?

File: src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs [hunk 3]
```diff
@@ -174,7 +171,7 @@ private static Section CreateSectionHelper(short maxValue, short priorMask, shor
 #if DEBUG
             int maskCheck = CreateMaskFromHighValue(maxValue);
             int offsetCheck = priorOffset + CountBitsSet(priorMask);
-            Debug.Assert(maskCheck <= short.MaxValue && offsetCheck < 32, "Overflow on BitVector32");
+            Debug.Assert(maskCheck <= short.MaxValue, "Overflow on BitVector32");
```
Comment (stephentoub) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs]: Why did you remove the second part of the assert?

File: src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs [hunk 4]
```diff
@@ -43,7 +43,7 @@ public ListDictionary(IComparer comparer)
                     while (node != null)
                     {
                         object oldKey = node.key;
-                        if (oldKey != null && oldKey.Equals(key))
+                        if (oldKey.Equals(key))
```
Comment (stephentoub) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs]: What guarantees that ```oldKey``` is not null?  Is this because nulls can't be put into the collection as keys to begin with?

File: src/System.Collections.Specialized/src/System/Collections/Specialized/NameObjectCollectionBase.cs [hunk 5]
```diff
@@ -591,7 +591,7 @@ void ICollection.CopyTo(Array array, int index)
 
                 if (index < 0)
                 {
-                    throw new ArgumentOutOfRangeException("index", SR.Format(SR.IndexOutOfRange, index.ToString(CultureInfo.CurrentCulture)));
+                    throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_NeedNonNegNum);
```
Comment (stephentoub) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/NameObjectCollectionBase.cs]: It seems the original intention in both of these cases was to include the value in the exception.  So shouldn't these instead be:
```
throw new ArgumentOutOfRangeException("index", index, SR.ArgumentOutOfRange_NeedNonNegNum);
```
?

File: src/System.Collections.Specialized/src/System/Collections/Specialized/OrderedDictionary.cs [hunk 6]
```diff
@@ -46,11 +46,6 @@ public OrderedDictionary(int capacity, IEqualityComparer comparer)
 
         private OrderedDictionary(OrderedDictionary dictionary)
         {
-            if (dictionary == null)
-            {
-                throw new ArgumentNullException("dictionary");
-            }
-
```
Comment (stephentoub) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/OrderedDictionary.cs]: I suggest replacing this with an ```Debug.Assert(dictionary != null);```

File: src/System.Collections.Specialized/tests/BitVector32/CreateSectionShortSectionTests.cs [hunk 7]
```diff
@@ -159,5 +156,35 @@ public void Test01()
...
+
+        [Fact]
+        public void Set_SectionInt_OutOfBoundsInt()
+        {
+            BitVector32.Section s1 = BitVector32.CreateSection(short.MaxValue);
+            BitVector32.Section s2 = BitVector32.CreateSection(short.MaxValue, s1);
+
+            BitVector32 bv = new BitVector32();
+
+            // This throws in Debug, but not in Release.
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/BitVector32/CreateSectionShortSectionTests.cs]: Why?

File: src/System.Collections.Specialized/tests/HybridDictionary/SetItemObjObjTests.cs [hunk 8]
```diff
@@ -647,5 +644,17 @@ public void Test01()
                 Assert.False(true, string.Format("Error, failed to set to null"));
             }
         }
+
+        [Fact]
+        public void Set_CaseInsensitive_Works()
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/HybridDictionary/SetItemObjObjTests.cs]: There wasn't already a test for this?

File: src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs [hunk 9]
```diff
@@ -96,7 +93,10 @@ public void Test01()
             {
                 Assert.False(true, string.Format("Error, returned Keys.Count = {0}", ks.Count));
             }
-            arr = Array.CreateInstance(typeof(Object), len);
+
+            Assert.False(ks.IsSynchronized);
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs]: What does this have to do with this test?  I realize these tests are already a mess, but for anything new we add, we should strive to do so in a manner that fits as much as possible with how we'd like to the tests to be.

File: src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs [hunk 10]
```diff
@@ -285,6 +286,16 @@ public void Test01()
...
+
+            // Modifying the dictionary makes MoveNext and Reset throw
+            ks = Hd.Keys;
+            IEnumerator e = ks.GetEnumerator();
+            Assert.True(e.MoveNext());
+            Hd.Add("New_Key", "Value");
+
+            Assert.NotNull(e.Current);
+            Assert.Throws<InvalidOperationException>(() => e.MoveNext());
+            Assert.Throws<InvalidOperationException>(() => e.Reset());
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs]: Should these be in their own, appropriately-named test method?

File: src/System.Collections.Specialized/tests/NameObjectCollectionBase/KeysCollectionCopyToTests.cs [hunk 11]
```diff
@@ -258,9 +255,9 @@ void CheckArray(Array array, ArrayList expected)
             }
             for (int i = 0; i < expected.Count; i++)
             {
-                if (((String[])array)[i] != (String)expected[i])
+                if (((string[])array)[i] != (string)expected[i])
                 {
-                    Assert.False(true, string.Format("Value {0} is incorrect.  array[{0}]={1}, should be {2}", i, ((String[])array)[i], (String)expected[i]));
+                    Assert.False(true, string.Format("Value {0} is incorrect.  array[{0}]={1}, should be {2}", i, ((string[])array)[i], (string)expected[i]));
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/NameObjectCollectionBase/KeysCollectionCopyToTests.cs]: As long as you're changing these lines, shouldn't this just be an Assert.NotSame check?

File: src/System.Collections.Specialized/tests/NameObjectCollectionBase/SetObjectTests.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        public void TestSetIndexValue()
+        {
+            var noc = new MyNameObjectCollection();
+            for (int i = 0; i < 10; i++)
+            {
+                var foo = new Foo();
+                noc.Add("Name_" + i, foo);
+
+                noc[i] = new Foo();
+            }
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/NameObjectCollectionBase/SetObjectTests.cs]: I don't see any asserts here, and the name of the test isn't particularly explanatory.  What is this testing?

File: src/System.Collections.Specialized/tests/NameValueCollection/CtorIntIKCTests.cs [hunk 13]
```diff
@@ -152,7 +149,7 @@ public void Test01()
                 {
                     Assert.False(true, string.Format("Error, returned null instead of {0} ", exp3));
                 }
-                else if (String.Compare(nvc["Everything Is Equal!"], exp3) != 0)
+                else if (string.Compare(nvc["Everything Is Equal!"], exp3) != 0)
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/NameValueCollection/CtorIntIKCTests.cs]: The vast majority of the changes in this PR are search-and-replace / refactorings to change type names into keywords, e.g. ```String``` to ```string```.  In the future, if you're going to make such changes (we've actually shied away from having such changes done as we plan to do a full sweep for such things programmatically at some point), please do them in a separate commit, as doing so in a separate commit will make it easier to review both these programmatic changes and the few others that aren't.  Thanks.

File: src/System.Collections.Specialized/tests/StringCollection/AddStrTests.cs [hunk 14]
```diff
@@ -170,11 +167,24 @@ public void Test01()
...
+        public void TestAddIList()
+        {
+            IList list = new StringCollection();
+            for (int i = 0; i < 10; i++)
+            {
+                list.Add("String_" + i);
+                Assert.True(list.Contains("String_" + i));
+            }
+
+            Assert.Equal(10, list.Count);
```
Comment (stephentoub) [inline src/System.Collections.Specialized/tests/StringCollection/AddStrTests.cs]: You could move this line into the loop at the end and change it to be:
```C#
Assert.Equal(i + 1, list.Count);
```

**Comment by stephentoub** [comment]
Thanks for the changes, @hughbe.  I left some comments, mostly about separating out some of the changes into separate commits (in particular separating out the product changes from the test changes).

**Comment by hughbe** [comment]
@stephentoub all done!

File: src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs [hunk 15]
```diff
@@ -66,10 +66,7 @@ public BitVector32(BitVector32 value)
             set
             {
 #if DEBUG
```
Comment (ianhays) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs]: This preprocessor directive is no longer necessary

File: src/System.Collections.Specialized/tests/HybridDictionary/SetItemObjObjTests.cs (hunk 8 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/HybridDictionary/SetItemObjObjTests.cs]: Not for setting a key on a case insensitive hybrid dictionary where the key to set is the same case as the key added.

**Comment by stephentoub** [comment]
> all done!

Thanks, but there are still a couple of questions I asked without responses... could you take a look?

File: src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs]: Yes, this is the case (please correct me if wrong)!

File: src/System.Collections.Specialized/src/Resources/Strings.resx (hunk 2 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/Resources/Strings.resx]: Done

File: src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs]: It is covered in the lines following it - this throws a InvalidOperationException if I recall correctly

Thus this was an improper assert and resulted in different behaviour in release and debug

File: src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs]: Also, I've investigated the code and key is never null IMO

File: src/System.Collections.Specialized/src/System/Collections/Specialized/NameObjectCollectionBase.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/NameObjectCollectionBase.cs]: Done

File: src/System.Collections.Specialized/src/System/Collections/Specialized/OrderedDictionary.cs (hunk 6 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/OrderedDictionary.cs]: Done

File: src/System.Collections.Specialized/tests/BitVector32/CreateSectionShortSectionTests.cs (hunk 7 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/BitVector32/CreateSectionShortSectionTests.cs]: There is a Debug.Assert present in the code for setting an object by index

File: src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs]: Done

File: src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs (hunk 15 above)
Comment (hughbe) [inline src/System.Collections.Specialized/src/System/Collections/Specialized/BitVector32.cs]: Will remove when I consolidate PR feedback

File: src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs (hunk 10 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs]: Done

File: src/System.Collections.Specialized/tests/NameObjectCollectionBase/KeysCollectionCopyToTests.cs (hunk 11 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/NameObjectCollectionBase/KeysCollectionCopyToTests.cs]: Done

File: src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs (hunk 9 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/ListDictionary/GetKeysTests.cs]: Done

File: src/System.Collections.Specialized/tests/NameObjectCollectionBase/SetObjectTests.cs (hunk 12 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/NameObjectCollectionBase/SetObjectTests.cs]: Fixed

File: src/System.Collections.Specialized/tests/NameValueCollection/CtorIntIKCTests.cs (hunk 13 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/NameValueCollection/CtorIntIKCTests.cs]: Done

File: src/System.Collections.Specialized/tests/StringCollection/AddStrTests.cs (hunk 14 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/StringCollection/AddStrTests.cs]: Done

**Comment by hughbe** [comment]
@stephentoub
My bad! I've answered your questions - i think they're all still there after rebasing, so I've responded to all your comments
Sorry!

**Comment by Clockwork-Muse** [comment]
@ianhays #2920 - rejected as "won't fix, not important enough". 
@hughbe - I _strongly_ recommend you look at the work I've already done.  Much of the contents of the original tests is outright noise.  A bunch of stuff is already "done" (including BitVector32, to which I added relevant tests - I need to break up my PRs more).  A few things need to be checked, mostly relating to string comparisons (I believe I assumed the wrong thing about how they interacted with culture in a couple of places).

**Comment by hughbe** [comment]
@clockwork-muse
Sure thing I'll take a look later 
What I intended by this PR was not a cleanup of the tests, but to add missing tests

**Comment by hughbe** [comment]
Oh and don't worry, Im doing my best not to clash with what you've spent time doing - that wouldn't be good for any of us ;)

**Comment by Clockwork-Muse** [comment]
...well, I meant you could steal and submit it, but if your plan was to avoid such clashes I'll go ahead and (break them up and) submit them...

**Comment by hughbe** [comment]
Oh right, whoops, sorry for misunderstanding

**Comment by stephentoub** [comment]
A few more nits, but otherwise LGTM.  Thanks for the clean up and the additional tests.

**Comment by hughbe** [comment]
Comments addressed, and PR rebase - happy to help!

**Comment by stephentoub** [comment]
LGTM.  Thanks, @hughbe.
